### PR TITLE
Feature/progress bar

### PR DESCRIPTION
**Pull request by edwinmp**
Release 0.1.0

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        }
+    });
+    
+    grunt.loadNpmTasks("grunt-contrib-compress");
+    grunt.loadNpmTasks("grunt-contrib-clean");
+    grunt.loadNpmTasks("grunt-contrib-watch");
+    grunt.loadNpmTasks("grunt-contrib-copy");
+    grunt.loadNpmTasks("grunt-webpack");
+
+    grunt.registerTask("default", [ "clean build", "watch" ]);
```
Comment (Andries-Smit) [inline Gruntfile.js]: Please add `grunt-check-dependencies` like [https://github.com/akileng56/Range-slider/blob/master/Gruntfile.js](https://github.com/akileng56/Range-slider/blob/master/Gruntfile.js)

File: README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# Progress bar
+Displays progress bar
+
+## Features
+* Show percentage of progress based on value
+* Render different bar types that is; default, striped and animated
```
Comment (Andries-Smit) [inline README.md]: Rephrase? 

> Render bar types; plane, striped or animated strips.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+* Show different bar colors that is; success, info, warning, danger
```
Comment (Andries-Smit) [inline README.md]: Rephrase?
> Bar colours; success, info, warning or danger

File: README.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+## Dependencies
+Mendix 6
+
+## Demo project
+https://progressbar-demo.mxapps.io/
+
+## Issues
+Please report issues at https://github.com/mendixlabs/progress-bar/issues.
+
```
Comment (Andries-Smit) [inline README.md]: 3 newline and the end, should 1

File: karma.conf.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        concurrency: Infinity,
+        coverageReporter: {
+            dir: "./dist/testresults",
+            reporters: [
+                { type: "json", subdir: ".", file: "coverage.json" },
+                { type: "text" }
+            ]
+        }
+    });
+};
```
Comment (Andries-Smit) [inline karma.conf.js]: Missing newline

File: package.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
+{
+  "name": "progressbar",
+  "version": "0.1.0",
+  "description": "Progress bar used to show percentage of progress for values 0 to 100",
```
Comment (Andries-Smit) [inline package.json]: `for values 0 to 100` is not true anymore. So it could be removed.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    "pretest": "tsc",
+    "test": "karma start --single-run --codeCoverage && node tests/remap.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": ""
+  },
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": ""
```
Comment (Andries-Smit) [inline package.json]: Missing URL

File: package.json [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  "directories": {
+    "test": "test"
+  },
+  "scripts": {
+    "pretest": "tsc",
+    "test": "karma start --single-run --codeCoverage && node tests/remap.js"
+  },
+  "repository": {
+    "type": "git",
+    "url": ""
```
Comment (Andries-Smit) [inline package.json]: missing URL

File: package.json [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  },
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": ""
+  },
+  "devDependencies": {
```
Comment (Andries-Smit) [inline package.json]: Please update (dev)dependecies with `> npm-check`

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    render() {
+        const { barType, bootstrapStyle, colorSwitch, maximumValue, onClickMicroflow, progress } = this.props;
+        const percentage = this.progressValue(progress, maximumValue);
+        return DOM.div({ className: "widget-progressbar" },
+            DOM.div(
+                {
+                    className: classNames("progress", {
+                        "red-progressbar-text": maximumValue < 1,
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: It not good to explicit name colours in styles. Better use `alert-progressbar-text`

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        }
+
+        return this.calculatePercentage(progress, maximumValue);
+    }
+
+    private calculatePercentage(progress: number, maxValue: number) {
+        return Math.round((progress / maxValue) * 100);
+    }
+
+    private executeMicroflow (action: string | undefined, guid: string | undefined) {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Suggestion API of progress circle also support open page.

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+    private getProgressBarProps(): ProgressBarProps {
+        const { bootstrapStyleAttribute, contextObject, maximumValueAttribute, progressAttribute } = this;
+        let progress = 0;
+        let bootstrapStyle = this.progressStyle;
+        let maximumValue = 100;
+        if (this.contextObject) {
+            progress = Math.round(parseInt(contextObject.get(progressAttribute) as string, 10));
+            bootstrapStyle = contextObject.get(bootstrapStyleAttribute || "") as BootstrapStyle || bootstrapStyle;
+            maximumValue = maximumValueAttribute ? Number(contextObject.get(maximumValueAttribute)) : maximumValue;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.ts]: use `as number`

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    private getProgressBarProps(): ProgressBarProps {
+        const { bootstrapStyleAttribute, contextObject, maximumValueAttribute, progressAttribute } = this;
+        let progress = 0;
+        let bootstrapStyle = this.progressStyle;
+        let maximumValue = 100;
+        if (this.contextObject) {
+            progress = Math.round(parseInt(contextObject.get(progressAttribute) as string, 10));
+            bootstrapStyle = contextObject.get(bootstrapStyleAttribute || "") as BootstrapStyle || bootstrapStyle;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.ts]: dont get attribute "". `contextObject.get(bootstrapStyleAttribute || "")`

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+    private resetSubscriptions() {
+        this.unsubscribeAll();
+        if (this.contextObject) {
+            this.subscribe({
+                callback: () => this.updateRendering(),
+                guid: this.contextObject.getGuid()
+            });
+
+            [ this.progressAttribute, this.bootstrapStyleAttribute, this.maximumValueAttribute ].forEach((attribute) =>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.ts]: 👍 like it!

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                    callback: () => this.updateRendering(),
+                    guid: this.contextObject.getGuid()
+                })
+            );
+        }
+    }
+}
+
+// tslint:disable : only-arrow-functions
+dojoDeclare("com.mendix.widget.ProgressBar.ProgressBar", [ WidgetBase ], function (Source: any) {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.ts]: no need for a space between `function (Source: any)`

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    <icon>
+        iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJ bWFnZVJlYWR5ccllPAAAAKdJREFUeNpi/P//PwMlgImBQjDwBrDw8fFF/vnzR44szSwsj1iUlZUl 161b10GOAUFBQcVMkpKSf1hZWRl27NgBxvjY6Bikl4WHh+c/Gxsbw4cPH8Cm4mOjA5BekAH/2NnZ Gd68eQMWxMfGYsA/Fi4uLrABb9++hSvExUYHIL0snJyc/0H+mTt3LlwCFxsdgPSyAE1+2tDQkEFO LAD1PmMczQsMAAEGACNuS4UbtbU3AAAAAElFTkSuQmCC
+    </icon>
+    <properties>
+        <property key="progressAttribute" type="attribute">
+            <caption>Progress attribute</caption>
+            <category>Data source</category>
+            <description>Holds the progress value</description>
+            <attributeTypes>
+                <attributeType name="Integer" />
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Missing types `long` and `decimal`

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+            </attributeTypes>
+        </property>
+        <property key="maximumValueAttribute" type="attribute" required="false">
+            <caption>Maximum value attribute</caption>
+            <category>Data source</category>
+            <description>Highest progress value. If not set, maximum value defaults to 100</description>
+            <attributeTypes>
+                <attributeType name="Integer" />
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Missing types `long` and `decimal`

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+            <caption>Colour switch value</caption>
+            <category>Display</category>
+            <description>The percentage at which the text gets a contrast colour (class) from black to white</description>
+        </property>
+        <property key="barType" type="enumeration" required="true" defaultValue="default">
+            <caption>Type</caption>
+            <category>Display</category>
+            <description>Render the type of progress bar: default, striped or animated </description>
+            <enumerationValues>
+                <enumerationValue key="default">Default</enumerationValue>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.xml]: `Default` does not represent the real rendering. `Plain` is maybe better, open to suggestions.

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+            <category>Behaviour</category>
+            <description>Microflow to call when the progress bar is clicked</description>
+            <returnType type="Void" />
+        </property>
+        <property key="progressStyle" type="enumeration" required="true" defaultValue="default">
+            <caption>Progress style</caption>
+            <category>Display</category>
+            <description>The color of the progress bar using bootstrap classes. Only used when 'Style attribute' is not set</description>
+            <enumerationValues>
+                <enumerationValue key="default">Default (None)</enumerationValue>
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ProgressBar.xml]: `Default (None)` => `None`

File: src/com/mendix/widget/ProgressBar/components/Alert.ts [hunk 20]
```diff
@@ -0,0 +1,5 @@
+import { DOM, StatelessComponent } from "react";
+
+export const Alert: StatelessComponent<{ message: string }> = (props) =>
+    DOM.div({ className: "alert alert-danger widget-validation-alert" }, props.message);
+
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/Alert.ts]: missing newline

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+                    {
+                        className: classNames("progress-bar", `progress-bar-${bootstrapStyle}`, {
+                            "active": barType === "animated",
+                            "progress-bar-striped": barType === "striped" || barType === "animated",
+                            "widget-progressbar-negative": percentage < 0
+                        }),
+                        style: { width: `${Math.abs(percentage)}%` }
+                    },
+                    progress
+                        ? maximumValue < 1 ? "Invalid" : `${this.calculatePercentage(progress, maximumValue)}%`
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: nested ternary operators are hard to read;
Add () around inner statement 
Or use if statements

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+            ),
+            this.state.alertMessage ? createElement(Alert, { message: this.state.alertMessage }) : null
+        );
+    }
+
+    private progressValue(progress: number | null, maximumValue: number) {
+        if (typeof progress !== "number") {
+            return 0;
+        } else if (maximumValue < 1) {
+            window.console.warn("The maximum value is less than one. Progress is set to Invalid");
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: console error should not be used. UI messages should be shown to the user directly via an alert

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+                            "widget-progressbar-negative": percentage < 0
+                        }),
+                        style: { width: `${Math.abs(percentage)}%` }
+                    },
+                    progress
+                        ? maximumValue < 1 ? "Invalid" : `${this.calculatePercentage(progress, maximumValue)}%`
+                        : ""
+                )
+            ),
+            this.state.alertMessage ? createElement(Alert, { message: this.state.alertMessage }) : null
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Can we move the statement to the component?

File: src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+    });
+
+    it("should render the progress", () => {
+        const wrapper = renderWrapper({ maximumValue, progress: 20 });
+
+        expect(wrapper.childAt(0).text()).toEqual(`${20}%`);
+        wrapper.setProps({ maximumValue, progress: -20 });
+        expect(wrapper.childAt(0).text()).toEqual(`${-20}%`);
+        wrapper.setProps({ maximumValue, progress: 300 });
+        expect(wrapper.childAt(0).text()).toEqual(`${300}%`);
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts]: split into 3 different tests

File: src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts [hunk 25]
```diff
@@ -0,0 +1,167 @@
...
+        expect(progressbar.text()).toEqual("Invalid");
+    });
+
+    it("should not render the progress label when no progress value is specified", () => {
+        const progressbar = getProgressbar({ maximumValue, progress: null }).childAt(0);
+
+        expect(progressbar.text()).toEqual("");
+    });
+
+    it("should not have the class widget-progressbar-text-contrast when progress exceeds the threshold", () => {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts]: add opposite test too.

File: src/com/mendix/widget/ProgressBar/ui/ProgressBar.css [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+.widget-progressbar-clickable:hover {
+    cursor: pointer;
+}
+
+.widget-progressbar-negative {
+    float: right;
+    display: block
+}
+
+.widget-progressbar-text-contrast.red-progressbar-text .progress-bar {
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/ui/ProgressBar.css]: don't use colour names in classes, use functional names

File: tests/helpers/structureMatcher.ts [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+            }
+        }
+
+        return { type, props, children: children.filter(c => c != null).map(reactToStructure) };
+    }
+}
+
+function isJasmineAny(obj: any): obj is jasmine.Any {
+    return obj != null && typeof obj.asymmetricMatch === "function";
+}
```
Comment (Andries-Smit) [inline tests/helpers/structureMatcher.ts]: missing newline

File: tests/mocks/Mendix.ts [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+        },
+        scope?: any
+    ): void { /* */ }
+    showLogin(messageCode: number): void { /* */ }
+}
+
+let mxMockObject =  MxMock.prototype;
+mxMockObject.ui = MxUiMock.prototype;
+
+export const mockMendix = mxMockObject;
```
Comment (Andries-Smit) [inline tests/mocks/Mendix.ts]: missing newline

File: tslint.json [hunk 29]
```diff
@@ -0,0 +1,94 @@
...
+        ],
+        "curly": false,
+        "max-line-length": [ true, 120 ]
+    },
+    "jsRules": {
+        "trailing-comma": [ false ],
+        "object-literal-sort-keys": false,
+        "array-bracket-spacing": [ true, "always" ]
+    }
+}
```
Comment (Andries-Smit) [inline tslint.json]: missing newline

File: webpack.config.js [hunk 30]
```diff
@@ -0,0 +1,45 @@
...
+        }
+    },
+    errorDetails: true,
+    module: {
+        loaders: [
+            { test: /\.ts$/, loader: "ts-loader" },
+            { test: /\.json$/, loader: "json" },
+            { test: /\.css$/, loader: ExtractTextPlugin.extract("style-loader", "css-loader") }
+        ],
+        postLoaders: [ {
```
Comment (Andries-Smit) [inline webpack.config.js]: postloader should be conditional in the kara config

**Review by Andries-Smit — COMMENTED** [review]
please update

File: Gruntfile.js (hunk 1 above)
Comment (edwinmp) [reply Gruntfile.js]: Done.

File: README.md (hunk 2 above)
Comment (edwinmp) [reply README.md]: Done.

File: README.md (hunk 3 above)
Comment (edwinmp) [reply README.md]: Done.

File: README.md (hunk 4 above)
Comment (edwinmp) [reply README.md]: Removed.

File: karma.conf.js (hunk 5 above)
Comment (edwinmp) [reply karma.conf.js]: Added.

File: package.json (hunk 6 above)
Comment (edwinmp) [reply package.json]: Removed.

File: package.json (hunk 8 above)
Comment (edwinmp) [reply package.json]: Added.

File: package.json (hunk 7 above)
Comment (edwinmp) [reply package.json]: Added.

File: package.json (hunk 9 above)
Comment (edwinmp) [reply package.json]: Done.

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts (hunk 10 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Done.

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts (hunk 21 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Fixed.

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts (hunk 23 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Not without having to deal with type errors given the recent strictness. They don't like components returning null. Frankly, I find it better & cleaner this way :)

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts (hunk 22 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Removed.

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 13 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: If bootstrapStypeAttribute is not set, it passes in "" which lets the function return undefined/null.
It's a shortcut & more accurate way of writing
```
bootstrapStyle = boostrapStypeAttribute
    ? contextObject.get(bootstrapAttribute) as BootstrapStyle
    : bootstrapStyle
```

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 12 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: Done.

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 14 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: :)

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 15 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: Removed

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml (hunk 16 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Done.

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml (hunk 17 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Done.

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml (hunk 19 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Done.

File: src/com/mendix/widget/ProgressBar/components/Alert.ts (hunk 20 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/Alert.ts]: Added.

File: src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts (hunk 24 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts]: Done.

File: src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts (hunk 25 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/__tests__/ProgressBar.spec.ts]: Done.

File: src/com/mendix/widget/ProgressBar/ui/ProgressBar.css (hunk 26 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ui/ProgressBar.css]: Changed to `widget-progressbar-alert`

File: tests/helpers/structureMatcher.ts (hunk 27 above)
Comment (edwinmp) [reply tests/helpers/structureMatcher.ts]: Added.

File: tests/mocks/Mendix.ts (hunk 28 above)
Comment (edwinmp) [reply tests/mocks/Mendix.ts]: Added.

File: tslint.json (hunk 29 above)
Comment (edwinmp) [reply tslint.json]: Added.

File: webpack.config.js (hunk 30 above)
Comment (edwinmp) [reply webpack.config.js]: Moved.

File: src/com/mendix/widget/ProgressBar/components/ProgressBar.ts (hunk 11 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/ProgressBar.ts]: Added page support.

File: karma.conf.js [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+        "react/lib/ExecutionEnvironment",
+        "react/lib/ReactContext",
+        "react/addons",
+        "jsdom"
+    ])
+});
+
+module.exports = function(config) {
+    if (config.codeCoverage) {
+        console.log("With instrumenter");
```
Comment (Andries-Smit) [inline karma.conf.js]: Remove the `console.log`

File: karma.conf.js (hunk 5 above)
Comment (Andries-Smit) [reply karma.conf.js]: Not really

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 13 above)
Comment (Andries-Smit) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: Don`t agree, its about `contextObject.get("")` that might cause issue later in live. You should get an error via the mx API as the attribute "" does not exist on the context object.

File: src/com/mendix/widget/ProgressBar/ProgressBar.xml (hunk 17 above)
Comment (Andries-Smit) [reply src/com/mendix/widget/ProgressBar/ProgressBar.xml]: Did you updated the Mendix Demo project too?

File: src/com/mendix/widget/ProgressBar/components/Alert.ts [hunk 32]
```diff
@@ -0,0 +1,6 @@
+import { DOM, StatelessComponent } from "react";
+
+export const Alert: StatelessComponent<{ message?: string }> = (props) =>
+    props.message
+        ? DOM.div({ className: "alert alert-danger widget-progressbar-alert" }, props.message)
+        : null as any;
```
Comment (Andries-Smit) [inline src/com/mendix/widget/ProgressBar/components/Alert.ts]: Remove `as any` update typing to
"@types/react": "^15.0.6",
"@types/react-dom": "^0.14.22",

**Review by Andries-Smit — CHANGES_REQUESTED** [review]
Last review.

File: karma.conf.js (hunk 5 above)
Comment (edwinmp) [reply karma.conf.js]: Sneaky bugger... added this time... I think. :)

File: src/com/mendix/widget/ProgressBar/ProgressBar.ts (hunk 13 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/ProgressBar.ts]: Reverted.

File: karma.conf.js (hunk 31 above)
Comment (edwinmp) [reply karma.conf.js]: Done.

File: src/com/mendix/widget/ProgressBar/components/Alert.ts (hunk 32 above)
Comment (edwinmp) [reply src/com/mendix/widget/ProgressBar/components/Alert.ts]: Done.

**Pull request closed without merging**